lib/charms/mongodb/v1/shards_interface.py

```diff
@@ -541,28 +558,96 @@ def _handle_changed_secrets(self, event) -> None:
+        return (key_file_contents is not None, tls_ca is not None)
+
+    def update_member_auth(self, event, membership_auth) -> None:
+        """Updates the shard to have the same membership auth as the config-server."""
+        (key_file_contents, cluster_tls_ca) = membership_auth
+        tls_integrated = self.charm.model.get_relation(Config.TLS.TLS_PEER_RELATION)
+
+        # Edge case: shard has TLS enabled before having connected to the config-server. For TLS in
+        # sharded MongoDB clusters it is necessary that the subject and organisation name are the
+        # same in their CSRs. Re-requesting a cert after integrated with the config-server
+        # regenerates the cert with the appropriate configurations needed for sharding.
+        if cluster_tls_ca and tls_integrated and not self.has_requested_cluster_certs():
+            logger.info("Cluster implements internal membership auth via certificates")
+            self.charm.tls.request_certificate(param=None, internal=True)
+            self.charm.tls.request_certificate(param=None, internal=False)
+        elif key_file_contents and not cluster_tls_ca and not tls_integrated:
+            logger.info("Cluster implements internal membership auth via keyFile")
+
+        # Copy over keyfile regardless of whether the cluster uses TLS or or KeyFile for internal
+        # membership authentication. If TLS is disabled on the cluster this enables the cluster to
+        # have the correct cluster KeyFile readily available.
+        key_file_contents = self.database_requires.fetch_relation_field(
+            event.relation.id, KEYFILE_KEY
+        )
+        self.update_keyfile(key_file_contents=key_file_contents)
+
+    # Future PR - status updating for inconsistencies with TLS (i.e. shard has TLS but
+    # config-server does not and vice versa or CA-mismatch)
+
+    def get_cluster_passwords(self, event) -> Tuple:
+        """Retrieves shared cluster passwords."""
+        operator_password = self.database_requires.fetch_relation_field(
+            event.relation.id, OPERATOR_PASSWORD_KEY
+        )
+        backup_password = self.database_requires.fetch_relation_field(
+            event.relation.id, BACKUP_PASSWORD_KEY
+        )
+        return (operator_password, backup_password)
+
+    def update_cluster_passwords(self, event, operator_password, backup_password):
+        """Updates shared cluster passwords."""
+        try:
+            self.update_password(
+                username=OperatorUser.get_username(), new_password=operator_password
+            )
+            self.update_password(BackupUser.get_username(), new_password=backup_password)
+        except RetryError:
+            self.charm.unit.status = BlockedStatus("Shard not added to config-server")
+            logger.error(
+                "Shard could not be added to config server, failed to set operator password."
+            )
```

Comment:
Could you explain this? it's not immediately clear to me from the `update_password` logic

Reply:
better?